File: src/bitpacking.cs
using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;

public static class UrlBitPack
{
    private const int BaseId = 30_000_000;

    // ======== Public API ========

    public static string EncodeToBase64UrlGzip(IReadOnlyList<(int Id, int Type)> items)
    {
        var raw = EncodeRawBitPacked(items);
        var gz = Gzip(raw);
        return ToBase64Url(gz);
    }

    public static List<(int Id, int Type)> DecodeFromBase64UrlGzip(string base64Url)
    {
        var gz = FromBase64Url(base64Url);
        var raw = Gunzip(gz);
        return DecodeRawBitPacked(raw);
    }

    // ======== Bit-tight encoding (Option B) ========

    private static byte[] EncodeRawBitPacked(IReadOnlyList<(int Id, int Type)> items)
    {
        // compute k
        int maxOffset = 0;
        foreach (var (Id, _) in items)
        {
            int off = Id - BaseId;
            if (off < 0) throw new ArgumentOutOfRangeException(nameof(items), "Id < BaseId");
            if (off > maxOffset) maxOffset = off;
        }
        int k = Math.Max(1, 32 - BitOperations.LeadingZeroCount((uint)maxOffset)); // ceil(log2(maxOffset+1))

        using var ms = new MemoryStream();

        // header: [version=1][k][count u16 BE]
        ms.WriteByte(1);
        ms.WriteByte((byte)k);
        Span<byte> hdr = stackalloc byte[2];
        BinaryPrimitives.WriteUInt16BigEndian(hdr, (ushort)items.Count);
        ms.Write(hdr);

        var bw = new BitWriter(ms);
        foreach (var (Id, Type) in items)
        {
            int off = Id - BaseId;
            bw.WriteBits((uint)off, k);
            bw.WriteBits((uint)(Type & 0b11), 2);
        }
        bw.FlushFinalByte();

        return ms.ToArray();
    }

    private static List<(int Id, int Type)> DecodeRawBitPacked(byte[] data)
    {
        if (data.Length < 4) throw new FormatException("Too short");
        if (data[0] != 1) throw new FormatException("Unsupported version");

        int k = data[1];
        if (k <= 0 || k > 30) throw new FormatException("Invalid k");

        ushort count = BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(2, 2));
        var items = new List<(int Id, int Type)>(count);

        using var ms = new MemoryStream(data, 4, data.Length - 4, writable: false);
        var br = new BitReader(ms);

        for (int i = 0; i < count; i++)
        {
            uint off = br.ReadBits(k);
            uint type = br.ReadBits(2);
            items.Add((BaseId + (int)off, (int)type));
        }

        return items;
    }

    // ======== Helpers: Bit I/O ========

    private sealed class BitWriter
    {
        private readonly Stream _s;
        private int _bitPos = 8; // when 8 => empty
        private byte _cur;

        public BitWriter(Stream s) => _s = s;

        // Writes 'bitCount' bits from 'value', MSB-first.
        public void WriteBits(uint value, int bitCount)
        {
            for (int i = bitCount - 1; i >= 0; i--)
            {
                int bit = (int)((value >> i) & 1);
                if (_bitPos == 8) { _cur = 0; _bitPos = 0; }
                _cur = (byte)((_cur << 1) | bit);
                _bitPos++;
                if (_bitPos == 8)
                {
                    _s.WriteByte(_cur);
                    _bitPos = 8;
                }
            }
        }

        public void FlushFinalByte()
        {
            if (_bitPos > 0 && _bitPos < 8)
            {
                _cur = (byte)(_cur << (8 - _bitPos));
                _s.WriteByte(_cur);
                _bitPos = 8;
            }
        }
    }

    private sealed class BitReader
    {
        private readonly Stream _s;
        private int _bitsLeft = 0;
        private int _cur = 0;

        public BitReader(Stream s) => _s = s;

        // Reads 'bitCount' bits, MSB-first, returns as uint.
        public uint ReadBits(int bitCount)
        {
            uint v = 0;
            for (int i = 0; i < bitCount; i++)
            {
                if (_bitsLeft == 0)
                {
                    int b = _s.ReadByte();
                    if (b < 0) throw new EndOfStreamException();
                    _cur = b;
                    _bitsLeft = 8;
                }
                int msb = (_cur & 0x80) != 0 ? 1 : 0;
                v = (v << 1) | (uint)msb;
                _cur = (_cur << 1) & 0xFF;
                _bitsLeft--;
            }
            return v;
        }
    }

    // ======== Helpers: gzip + base64url ========

    private static byte[] Gzip(byte[] input)
    {
        using var ms = new MemoryStream();
        using (var gz = new GZipStream(ms, CompressionLevel.SmallestSize, leaveOpen: true))
            gz.Write(input, 0, input.Length);
        return ms.ToArray();
    }

    private static byte[] Gunzip(byte[] input)
    {
        using var src = new MemoryStream(input);
        using var gz = new GZipStream(src, CompressionMode.Decompress);
        using var dst = new MemoryStream();
        gz.CopyTo(dst);
        return dst.ToArray();
    }

    private static string ToBase64Url(byte[] data)
        => Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private static byte[] FromBase64Url(string s)
    {
        string b64 = s.Replace('-', '+').Replace('_', '/');
        switch (b64.Length % 4) { case 2: b64 += "=="; break; case 3: b64 += "="; break; }
        return Convert.FromBase64String(b64);
    }
}

File: src/bitpacking.js
// node >=14
const zlib = require("zlib");

// ======== Public API ========

function encodeToBase64UrlGzip(items) {
  const raw = encodeRawBitPacked(items);
  const gz = zlib.gzipSync(raw, { level: zlib.constants.Z_BEST_COMPRESSION });
  return toBase64Url(gz);
}

function decodeFromBase64UrlGzip(token) {
  const gz = fromBase64Url(token);
  const raw = zlib.gunzipSync(gz);
  return decodeRawBitPacked(raw);
}

// ======== Bit-tight encoding (Option B) ========

const BASE_ID = 30_000_000;

function encodeRawBitPacked(items) {
  let maxOffset = 0;
  for (const { Id, Type } of items) {
    const off = Id - BASE_ID;
    if (off < 0) throw new Error("Id < BaseId");
    if (off > maxOffset) maxOffset = off;
  }
  let k = Math.max(1, 32 - Math.clz32(maxOffset >>> 0));

  // Header: [1][k][count u16 BE]
  const header = Buffer.alloc(4);
  header[0] = 1;
  header[1] = k;
  header.writeUInt16BE(items.length, 2);

  const bw = new BitWriter();
  for (const { Id, Type } of items) {
    const off = Id - BASE_ID;
    bw.writeBits(off >>> 0, k);
    bw.writeBits(Type & 0b11, 2);
  }
  const payload = bw.finish();

  return Buffer.concat([header, payload]);
}

function decodeRawBitPacked(buf) {
  if (buf.length < 4) throw new Error("Too short");
  if (buf[0] !== 1) throw new Error("Unsupported version");
  const k = buf[1];
  if (k <= 0 || k > 30) throw new Error("Invalid k");

  const count = buf.readUInt16BE(2);
  const br = new BitReader(buf.subarray(4));

  const items = [];
  for (let i = 0; i < count; i++) {
    const off = br.readBits(k) >>> 0;
    const type = br.readBits(2) >>> 0;
    items.push({ Id: BASE_ID + off, Type: type });
  }
  return items;
}

// ======== Helpers: Bit I/O ========

class BitWriter {
  constructor() {
    this._bytes = [];
    this._cur = 0;
    this._bits = 0; // number of bits currently in _cur (0..8)
  }
  writeBits(value, bitCount) {
    for (let i = bitCount - 1; i >= 0; i--) {
      const bit = (value >>> i) & 1;
      this._cur = (this._cur << 1) | bit;
      this._bits++;
      if (this._bits === 8) {
        this._bytes.push(this._cur);
        this._cur = 0;
        this._bits = 0;
      }
    }
  }
  finish() {
    if (this._bits > 0) {
      this._cur = this._cur << (8 - this._bits);
      this._bytes.push(this._cur & 0xff);
      this._cur = 0;
      this._bits = 0;
    }
    return Buffer.from(this._bytes);
  }
}

class BitReader {
  constructor(buf) {
    this._buf = buf;
    this._i = 0;
    this._cur = 0;
    this._bits = 0; // bits remaining in _cur (0..8)
  }
  readBits(bitCount) {
    let v = 0 >>> 0;
    for (let i = 0; i < bitCount; i++) {
      if (this._bits === 0) {
        if (this._i >= this._buf.length) throw new Error("Unexpected EOF");
        this._cur = this._buf[this._i++];
        this._bits = 8;
      }
      const msb = (this._cur & 0x80) ? 1 : 0;
      v = (v << 1) | msb;
      this._cur = (this._cur << 1) & 0xff;
      this._bits--;
    }
    return v >>> 0;
  }
}

// ======== Helpers: base64url ========

function toBase64Url(buf) {
  return buf.toString("base64").replace(/=+$/g, "").replace(/\+/g, "-").replace(/\//g, "_");
}
function fromBase64Url(s) {
  s = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = s.length % 4;
  if (pad === 2) s += "==";
  else if (pad === 3) s += "=";
  return Buffer.from(s, "base64");
}

module.exports = { encodeToBase64UrlGzip, decodeFromBase64UrlGzip };
